galaxy_ng/app/api/ui/serializers/synclist.py

```diff
@@ -28,13 +35,120 @@ class SyncListSerializer(serializers.ModelSerializer):
+                'Repository "{pulp_id}" not found while creating synclist'
+            raise ValidationError(
+                errmsg.format(pulp_id=repository_id))
+
+    def to_internal_value(self, data):
+        groups = data.get('groups', [])
+        data['groups'] = self._sanitize_groups(groups)
+
+        repository_data = data.get('repository', None)
+        if repository_data:
+            data['repository'] = self._get_repository(repository_data)
+
+        upstream_repository_data = data.get('upstream_repository', None)
+        if upstream_repository_data:
+            data['upstream_repository'] = self._get_repository(upstream_repository_data)
+        else:
+            # If not specified, use the default upstream repo
+            data['upstream_repository'] = \
+                AnsibleRepository.objects.get(name=default_repo_name)
+        return super().to_internal_value(data)
+
+    def _name_slug_from_group_names(self, groups_data):
+        ngs = auth_models.Group.objects.filter(name__in=[x.name for x in groups_data])
+        ngs = ngs.filter(name__ne=auth_models.RH_PARTNER_ENGINEER_GROUP)
+
+        if not ngs.filter():
+            # TODO: better error
+            errmsg = 'No Group found while creating synclist'
+            raise ValidationError(errmsg)
+
+        # FIXME: further filter for group name scope?
+
+        account_name = ngs.filter()[0].account_number()
+        name_slug = settings.GALAXY_API_SYNCLIST_NAME_FORMAT.format(account_name=account_name)
+
+        return name_slug, account_name
+
+    @transaction.atomic
+    def _create_repository(self, name_slug, account_name):
+        try:
+            description = f"Synclist repository for account {account_name}"
+
+            repository, created = AnsibleRepository.objects.get_or_create(name=name_slug)
+            repository.description = description
+            repository.save()
+            return repository
+        except Exception as exc:
+            log.exception(exc)
+            raise
+
+    def _update_distribution(self, name_slug, repository):
+        # Now the distro
+        try:
+            distribution, create = AnsibleDistribution.objects.get_or_create(
+                name=name_slug,
+                base_path=name_slug,
+            )
+        except Exception as exc:
+            log.exception(exc)
+            raise
+
+        # Now update the Distribution to point to the new Repository
+        distribution.repository = repository
+
+        distribution.save()
+        return distribution
+
     @transaction.atomic
     def create(self, validated_data):
         collections_data = validated_data.pop("collections")
 
         namespaces_data = validated_data.pop("namespaces")
+        users_data = validated_data.pop("users")
+        groups_data = validated_data.pop("groups")
+
+        repository = validated_data.pop("repository", None)
+
+        name_slug, account_name = self._name_slug_from_group_names(groups_data)
```

Comment:
We don't need to auto generate the sync list names from groups on the API. When creating synclists via the API the name, repository and distribution should all be set explicitly. The API is intended for direct use by a user and auto generating the names off of a set of obscure rules surrounding groups is going to be confusing.